dpcpp/factorization/par_ilut_select_kernels.hpp.inc

```diff
@@ -0,0 +1,444 @@
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+namespace kernel {
+
+
+constexpr auto searchtree_width = 1 << sampleselect_searchtree_height;
+constexpr auto searchtree_inner_size = searchtree_width - 1;
+constexpr auto searchtree_size = searchtree_width + searchtree_inner_size;
+
+constexpr auto sample_size = searchtree_width * sampleselect_oversampling;
+
+constexpr auto basecase_size = 1024;
+constexpr auto basecase_local_size = 4;
+constexpr auto basecase_block_size = basecase_size / basecase_local_size;
+
+
+// must be launched with one thread block and block size == searchtree_width
+/**
+ * @internal
+ *
+ * Samples `searchtree_width - 1` uniformly distributed elements
+ * and stores them in a binary search tree as splitters.
+ */
+template <typename ValueType, typename IndexType>
+void build_searchtree(const ValueType* __restrict__ input, IndexType size,
+                      remove_complex<ValueType>* __restrict__ tree_output,
+                      sycl::nd_item<3> item_ct1,
+                      remove_complex<ValueType>* sh_samples)
+{
+    using AbsType = remove_complex<ValueType>;
+    auto idx = item_ct1.get_local_id(2);
+    AbsType samples[sampleselect_oversampling];
+    // assuming rounding towards zero
+    // auto stride = remove_complex<ValueType>(size) / sample_size;
+#pragma unroll
+    for (int i = 0; i < sampleselect_oversampling; ++i) {
+        auto lidx = idx * sampleselect_oversampling + i;
+        auto val = input[static_cast<IndexType>(lidx * size / sample_size)];
```

Comment:
This may overflow for large inputs. better alternative would be switching between size < sample_size and size >= sample_size with different techniques (stride = size / sample_size or lidx * size / sample_size)

Reply:
I recover the stride computation, so it should not be the issue now?